Adjust ratio for opacity

jamesnw opened the pull request:
## Description
When foreground opacity is changed, the ratio is based on the premultiplied color. 
If the background opacity is changed, the ratio is based on both colors without any opacity.

## Related Issue(s)
#91 

## Steps to test/reproduce

mirisuzanne reviewed (approved):
looks good. Might be nice to have a checkerboard under the sample area, for when background color is not opaque?

jamesnw commented:
> looks good. Might be nice to have a checkerboard under the sample area, for when background color is not opaque?

@mirisuzanne Good idea. I added one, but is there a way to have a background color on a layer above an image? It looks like a color is only valid as the last value. I ended up using a linear-gradient to create a solid color, but there has to be a better way.

jamesnw commented:
@mirisuzanne or @stacyk I'm assigning to you, as I think there's potentially some style cleanup to prevent a layout jump when the warning message is shown, and also since the icon doesn't feel spaced correctly.

<img width="444" alt="image" src="https://github.com/user-attachments/assets/59756c15-33f0-4a9b-85b7-93bcd290663d" />

jgerigmeyer commented on src/lib/components/ratio/index.svelte:
```diff
@@ -1,15 +1,30 @@
+    if ($bg.alpha !== 1)
+      return 'Alpha is not considered when the background is not opaque.';
+    if ($fg.alpha !== 1)
+      return 'This ratio is our best estimate with transparency.';
```
"with transparency" seems a bit vague to me if users haven't read the full description. Maybe "given the specified foreground alpha value" or something?

jamesnw replied:
```suggestion
      return 'This ratio is our best estimate given the specified foreground alpha value.';
```

I like this suggestion, I'll change it.

jgerigmeyer commented on src/lib/components/ratio/index.svelte:
```diff
@@ -1,15 +1,30 @@
   let pass = $derived(ratio >= RATIOS.AA.Normal);
+  let alphaWarning = $derived.by(() => {
+    if ($bg.alpha !== 1)
+      return 'Alpha is not considered when the background is not opaque.';
```
Could we link these two notes directly to the fuller description (expanding if necessary)?

If not that, we could consider adding the fuller explanation in a tooltip/dialog here instead of in the expandable "issues" section.

jamesnw replied:
I think linking will be much easier once we're done with https://github.com/oddbird/oddcontrast/pull/223. I'm fine with moving it to a tooltip/dialog, but note we'd need to develop that pattern. I'll defer to @SondraE on what we should do.

jamesnw replied:
Correct- that would be great!

dvdherron commented:
> @mirisuzanne or @stacyk I'm assigning to you, as I think there's potentially some style cleanup to prevent a layout jump when the warning message is shown, and also since the icon doesn't feel spaced correctly.
> <img alt="image" width="444" src="https://private-user-images.githubusercontent.com/167908/394929825-59756c15-33f0-4a9b-85b7-93bcd290663d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc3MzQ3OTIsIm5iZiI6MTczNzczNDQ5MiwicGF0aCI6Ii8xNjc5MDgvMzk0OTI5ODI1LTU5NzU2YzE1LTMzZjAtNGE5Yi04NWI3LTkzYmNkMjkwNjYzZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyNFQxNjAxMzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05NjY2Zjg3MjlhZTRhNzQ4NzA1NjA1ZGY5MDZkMGNiMmViYzQwYjdmNDRlODVkODFiYTY2YTViN2E3NGQyOTVkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.nDI8qD-HvNecSeMhOz6Wefa4zOsRJanyaswd_d5oB5s">

I think I've fixed the layout jumping issues (in https://github.com/oddbird/oddcontrast/pull/222/commits/d4dccb85b4f2ea117a25ba79fe2016505a598d53) and cleaned up the warning message in (https://github.com/oddbird/oddcontrast/pull/222/commits/bf7d3eb9db0a9ab738cb20832f89a6a3b6e477e8)

@stacyk @SondraE Thoughts on how this is looking so far? Look at the warning message at the different break points and with passing and failing ratio.

SondraE reviewed (approved):
@dvdherron I think this is going in a good direction! 
There are a few breakpoints where the message seemed disconnected from the ratio and not aligned with the other content.

<img width="586" alt="Screenshot 2025-01-24 at 4 24 52 PM" src="https://github.com/user-attachments/assets/54f26922-5300-4aa2-b32b-5f90fbba5243" />
<img width="488" alt="Screenshot 2025-01-24 at 4 25 30 PM" src="https://github.com/user-attachments/assets/a06ad374-03a3-44b7-89ba-8d5b0f2dad12" />

mirisuzanne commented:
> > looks good. Might be nice to have a checkerboard under the sample area, for when background color is not opaque?
> 
> Good idea. I added one, but is there a way to have a background color on a layer above an image? It looks like a color is only valid as the last value. I ended up using a linear-gradient to create a solid color, but there has to be a better way.

@jamesnw - That's probably the best way to do it, yeah. I think it's a strange restriction on backgrounds, but it might pre-date transparent colors?

SondraE commented:
> I adjusted the spacing in some of those middle widths so that the warning isn't so far from the ratio.

This is looking better to me!
The only question I have is about the last image you posted that has no warning. Does there need to be so much space between the word _Ratio_ and the paragraph below it as well as between the rows of _Pass_ chiclets, when there isn't a warning?

stacyk commented:
> The only question I have is about the last image you posted that has no warning. Does there need to be so much space between the word _Ratio_ and the paragraph below it as well as between the rows of _Pass_ chiclets, when there isn't a warning?

I'm guessing the goal was to not have that text block jump around if you were editing the color. 

I wonder if that is something we should weigh against the gaps in the layout when no warning is shown. I realize this is a super difficult layout to adjust for since there are a few breakpoints with different layouts, then sometimes with a warning. I think that alpha will like be on or off, maybe toggled once so that jump doesn't bother me personally. I think I would prefer that jump over the gaps

what if, in this middle size we stack the heading and ratio result like it is at the larger sizes? Also, in the browser I adjusted the grid to the right ` .result-status` to include the following to align the content at the top: 
```
  grid-template-rows: min-content;
  padding-block-start: var(--shim-plus);
```

Middle sizes, with warning, heading spans full row
![middle-warning](https://github.com/user-attachments/assets/4f4e5c35-5fc0-4f57-8ff8-e2c3343f3e2a)

Middle sizes, no warning, heading spans full row
![middle-nowarning](https://github.com/user-attachments/assets/3672f4a4-d1ec-4a56-9f64-6dc38b42b864)


At the widest breakpoint, but the smallest size within, the warning box does grow to touch the paragraph. Pretty edge case but thought I would note it. 
![Screenshot 2025-01-28 at 2 14 10 PM](https://github.com/user-attachments/assets/c0bf953d-9612-4491-8517-b220daf00c79)

dvdherron commented:
> @dvdherron
> 
> > I adjusted the spacing in some of those middle widths so that the warning isn't so far from the ratio.
> 
> This is looking better to me! The only question I have is about the last image you posted that has no warning. Does there need to be so much space between the word _Ratio_ and the paragraph below it as well as between the rows of _Pass_ chiclets, when there isn't a warning?

Yeah, I was trying to hold space so that things wouldn't jump around when the warning pops up. I think what Stacy has sketched out below should work to close some of those gaps.

dvdherron commented:
@stacyk I implemented your suggestion last week and I think what I have done in https://github.com/oddbird/oddcontrast/pull/222/commits/52f355d0c81485eae058d6ea462c6304466c4cad removes a lot of the gaps made for holding space for the warning. 

Not sure if you and @SondraE should review now or wait until we sort out #223 and implement the possible changes to the warning message.

mirisuzanne commented:
![2025-02-03 12 11 32](https://github.com/user-attachments/assets/8265e34c-6427-49f4-b8a6-52e25c338f9b)

jamesnw commented:
With the color issues move merged in, I added links from the warnings to the color section. 

@dvdherron  The warnings will need to be styled. I assume the DOM will need to be changed somehow- feel free to do it, or let me know if you want me to.
<img width="427" alt="image" src="https://github.com/user-attachments/assets/8240bae7-bf4d-4960-bcc1-1f7a08f072ca" />

Because the issues are at the bottom of the page, the header won't be scrolled to the top of the page. I added a `:has(:target)` selector with a fairly heavy-handed highlight, which I think could be scaled back significantly. 

<img width="960" alt="image" src="https://github.com/user-attachments/assets/d5e7f83d-2ce6-4760-a66c-dcd02b8a2bb5" />

dvdherron commented:
> With the color issues move merged in, I added links from the warnings to the color section.
> 
> @dvdherron The warnings will need to be styled. I assume the DOM will need to be changed somehow- feel free to do it, or let me know if you want me to. 

Fixed the warnings in https://github.com/oddbird/oddcontrast/pull/222/commits/cf92e0831bd7d992a4683b5903f6e113b5e3179e
![bild](https://github.com/user-attachments/assets/35976fe3-11da-4126-b427-2e126f8ed222)



> 
> Because the issues are at the bottom of the page, the header won't be scrolled to the top of the page. I added a `:has(:target)` selector with a fairly heavy-handed highlight, which I think could be scaled back significantly.

Changed this to where just the heading is highlighted. Too subtle?

![bild](https://github.com/user-attachments/assets/71355c32-29bb-4429-9735-a404736b0376)

dvdherron commented:
> A gif of the layout bouncing as the ratio changes

Accidentally changed the size of the fixed width column for the ratio number. Fixed in https://github.com/oddbird/oddcontrast/pull/222/commits/607ea55a1a3f79d258d0f902a03e764bf95fc7ce

dvdherron commented:
@stacyk @SondraE @mirisuzanne I think this is ready for final review. 

The biggest change was stacking the ratio number on top of the eventual warning on the narrowest screens. Oherwise it felt like the warning box grew really tall:

![bild](https://github.com/user-attachments/assets/476e47da-58df-4a01-9ada-76ac08606677)

![bild](https://github.com/user-attachments/assets/9ef6036c-cbe9-47f4-9322-4bca22989caa)

And now that the color issues are linked to at the bottom of the page, the heading of the respective list item is subtly highlighted after following the "learn more" link in the warning. 

Is the color and the "diamond" too sublte?

![bild](https://github.com/user-attachments/assets/c82c4e19-4003-4b31-a24c-aa692a7523e9)

SondraE reviewed (approved):
This is all looking really good to me. I really like the idea of highlighting the title of the section the user is taken to after clicking Learn More. I wonder if it would feel even more related to the warning if the text color was the same pink as the warning text color and the same icon was used.

dvdherron commented:
> el even more related to the warning if the text color was the same pink as the warning text color and the same icon was used.

@SondraE Fixed in https://github.com/oddbird/oddcontrast/pull/222/commits/e129e616bf5bcfaed9f69b599ad42e56b480fc92

![image](https://github.com/user-attachments/assets/8f3c58b8-9919-4a52-8869-4846343193fc)

mirisuzanne commented on src/lib/components/ratio/index.svelte:
```diff
@@ -136,14 +202,23 @@
+    align-items: center;
     display: inline-flex;
     grid-area: number;
+    justify-content: flex-start;
```
I wonder if we want to remove this, and let it stretch to fill the space?

dvdherron replied:
removed in https://github.com/oddbird/oddcontrast/pull/222/commits/1044fb1934534a89fbd8ef3b56960ada12e791de

mirisuzanne commented on src/lib/components/ratio/index.svelte:
```diff
@@ -19,6 +41,15 @@
       <span class="sr-only">The contrast ratio is</span>
```
I know it's not new here, but I don't think this is useful screen reader context. The heading directly before this says 'current ratio' already.

dvdherron replied:
Agreed. On the topic, I wonder if (in another story) we should edit the click-to-copy buttons to say what exactly they are copying?

mirisuzanne replied:
Yeah, I don't know the best solution on those - but no harm oping an issue to explore it.

mirisuzanne commented on src/lib/components/ratio/index.svelte:
```diff
@@ -159,6 +234,11 @@
     text-align: center;
+
+    @include config.between('sm-page-break', 'lg-page-break') {
+      grid-template-rows: min-content;
```
I would use `align-content: start` here, if the goal is not to stretch rows so they fill the grid.

mirisuzanne commented on src/sass/config/scale/_ui.scss:
```diff
@@ -12,7 +12,7 @@ $swatch: 3.25rem;
 $icon-small: 0.65em;
 $icon-medium: 1.5em;
-$ratio-width: 10rem;
+$ratio-width: 8.5rem;
```
I'm not sure this needs to be a global token, when we only use it in one place and then dynamically override it. I think we could just use `--ratio-width` in the file where it's needed.

dvdherron replied:
moved closer to where it's used in https://github.com/oddbird/oddcontrast/pull/222/commits/1044fb1934534a89fbd8ef3b56960ada12e791de

mirisuzanne reviewed (commented):
this looks great. Just left a few notes on the code.

dvdherron commented on src/lib/components/ratio/index.svelte:
```diff
@@ -236,6 +238,7 @@
 
     @include config.between('sm-page-break', 'lg-page-break') {
+      align-content: start;
       grid-template-rows: min-content;
```
Got it. Removed in https://github.com/oddbird/oddcontrast/pull/222/commits/15b5251faa954a4aef9dc94dd6b65026779871f2

stacyk reviewed (approved):
Looks and works great!
